bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -641,6 +642,60 @@ public static boolean checkIsType(BType sourceType, BType targetType, List<TypeP
         }
     }
 
+    private static boolean checkIsType(Object sourceVal, BType sourceType, BType targetType,
```

Comment:
why we need another checkIsType method ? whats the purpose of this newly added method?

Reply:
The current `checkIsType` method is based solely on the source and target types. We check `checkIsLikeType` only if the entire source value is immutable.

But now with `readonly` fields, even though the entire source value is not immutable, due to the presence of `readonly` fields, an `is` check that fails for types, may pass for values. 

For example, from the sample in the description:

```ballerina
import ballerina/io;

const HUNDRED = 100;

type Foo record {
    HUNDRED i;
};

type Bar record {
    readonly string|int i;
};

public function main() {
    Bar b1 = {i: 123};
    io:println(<any> b1 is Foo); // prints `false`

    Bar b2 = {i: 100};
    io:println(<any> b2 is Foo); // prints `true` now, since the `i` field of `b2` is and will continue to be `100`
}
```

Here if we just use the current `checkIsType` with `Bar` and `Foo`, it will evaluate to false, since the source has `string|int` as the type for `i`, while the target has `HUNDRED` as the type.

But since the `i` field in the source is a `readonly` field, we know that the value for this field cannot change, so we look at the actual value to see if it belongs to the target type. So, in the case of `b1`, where `i` is `123`, it is still not a match, but for `b2` since `i` is `100`, and it can never change, the `is` check evaluates to `true`.

Will add a comment too.

Reply:
Added a comment in 24985b7.

Reply:
Added a comment in 24985b7.